### ISSUE TITLE
Remove unneeded entries from datacheckfps.csv

### ISSUE THREAD

**Issue by jteresco** [TravelMapping/DataProcessing#59]
Since VISIBLE_DISTANCE errors are no longer flagged for active routes, we have a lot of unmatched fps in unmatchedfps.log.  It would be good to automate a one-time cleanup of the fp list.

**Comment by yakra** [comment]
https://github.com/TravelMapping/HighwayData/pull/1657

**Comment by yakra** [comment]
I only removed VISIBLE_DISTANCE entries, and not all *unmatchedfps*.

**Comment by yakra** [comment]
That was, in the strictest sense, a "a one-time cleanup". :)
I can easily do that again, and it may be worthwhile...

There are now only 49 ([make that 48](https://github.com/TravelMapping/HighwayData/pull/1660), actually...) [unmatchedfps](http://tm.teresco.org/logs/unmatchedfps.log), for a *much* better signal-to-noise ratio.
Three of these are [nearmatchfps](http://tm.teresco.org/logs/nearmatchfps.log), and may be helpful for contributors correcting those. The rest, I suppose it may be better to leave to the authors of the respective regions to deal with as appropriate.
So, I erred on the side of caution and didn't cull these along with the VISIBLE_DISTANCE unmatchedfps.

More to the point, though, there are still 3139 VISIBLE_DISTANCE FPs marked in preview and devel systems, which may slowly filter into unmatchedfps.log as more systems go live (unless the author gets rid of them when activating the systems).

TL;DR, future cleanups may still be necessary. I guess I'll leave it to @jteresco whether to close or leave this issue open.

**Comment by jteresco** [comment]
I'll leave it open, since I think something more automated is a worthy future goal.

**Comment by michihdeu** [comment]
> More to the point, though, there are still 3139 VISIBLE_DISTANCE FPs marked in preview and devel systems

Out of more than 5,000 FPs in total...

**Comment by yakra** [comment]
How's this? :D
```C++
using namespace std;
#include <fstream>
#include <iostream>

class entry
{	public:
	string text;
	bool cull;
	entry *prev, *next;

	// init alpha
	entry() {prev = 0; next = 0; cull = 0;}
	// insertions
	entry(entry *cursor) // cursor, in this case, is a pointer to the prev entry
	{	prev = cursor;
		next = 0;
		cull = 0;
	}//*/
};

bool find(string needle, entry *haystack)
{	do {	if (needle == haystack->text)	return 1;
		haystack = haystack->next;
	   }
	while (haystack);
	return 0;
}

int main(int argc, char *argv[])
{	entry *cList = new entry;
	entry *fList = new entry;
	entry *cursor;
	if (argc != 4) {cout << "usage: ./fpcull CullFile FullFile OutputFile\n"; return 0;}
	// init ifstreams
	ifstream cFile(argv[1], ios::in);
	if (!cFile.is_open())	{ cout << argv[1] << " file not found!" << endl; return 0; }
	ifstream fFile(argv[2], ios::in);
	if (!fFile.is_open())	{ cout << argv[2] << " file not found!" << endl; return 0; }

	// create cull list
	cFile >> cList->text; //TODO enclose in an IF in case hell breaks loose (IE, no text in cFile)
	cursor = new entry(cList);
	while (cFile >> cursor->text)
	{	cursor->prev->next = cursor;
		cursor = new entry(cursor);
	}
	delete cursor;

	// create full list
	fFile >> fList->text; //TODO enclose in an IF in case hell breaks loose (IE, no text in cFile)
	cursor = new entry(fList);
	while (fFile >> cursor->text)
	{	cursor->prev->next = cursor;
		cursor = new entry(cursor);
	}
	delete cursor;

	// BADGER KULL
	cursor = fList;
	do {	if (find(cursor->text, cList)) cursor->cull = 1;
		cursor = cursor->next;
	   }
	while (cursor);

	// output
	ofstream oFile(argv[3]);
	cursor = fList;
	while (cursor)
	{	if (!cursor->cull)
			oFile << cursor->text << endl;
		cursor = cursor-> next;
	}
}```
It's not perfect; it choked on the line that Si removed from line 6613 of [this commit](https://github.com/TravelMapping/HighwayData/commit/1f30d1658ed95f053498c0db13787e23e8100509#diff-31b38d932857e7fd154f223fad0512c7), but it does a pretty bang-up job. Just requires a little manual checking.

**Comment by jteresco** [comment]
Looks better than the nonexistent one we had before!  Could you put in a pull request adding this to a directory like "CullDatacheckFPs" in this repository?

**Comment by yakra** [comment]
https://github.com/TravelMapping/HighwayData/commit/1f30d1658ed95f053498c0db13787e23e8100509#diff-31b38d932857e7fd154f223fad0512c7

**Comment by yakra** [comment]
> Could you put in a pull request adding this to a directory like "CullDatacheckFPs" in this repository?

#66

**Comment by yakra** [comment]
> It choked on [[a malformed line](https://github.com/TravelMapping/HighwayData/blob/d33d0f6e93eec8ce7f71f31af94336a473161e5f/datacheckfps.csv#L6613)] (because it expects every line to be one text string and every text string to be one line)

(Lines containing whitespace would have been split into multiple lines, instead of being left alone.)

Fixed in #124 -- and basically rewritten altogether. That earlier version was a bit of an embarrassment.

**Comment by yakra** [comment]
shell script to produce a list of `VISIBLE_DISTANCE` entries in active systems:
```sh
unmatchedlog="/home/yakra/TravelMapping/yakra/DataProcessing/siteupdate/cplusplus/logs/unmatchedfps.log"
repo="/home/yakra/TravelMapping/HighwayData"

cd $repo/hwy_data/_systems
CSVs=`ls *.csv | grep -v '_con\.csv'`

entries=`cat $unmatchedlog | grep VISIBLE_DISTANCE 2>&1`
for entry in $entries; do
    root=`echo $entry | cut -f1 -d';'`
    hs=`egrep "^.+;.+;.+;.*;.*;.*;$root;.*$" $CSVs | cut -f1 -d.`
    if [ `echo $hs | wc -w` -gt 0 ]; then
        level=`grep -v '^#' $repo/systems.csv | grep "^$hs;" | cut -f6 -d';' 2>&1`
        if [ `echo $level | wc -w` = '1' ]; then
            if [ $level = active ]; then
                echo $entry
            fi
        fi
    fi
done
```

**Comment by jteresco** [comment]
Still anything in this issue we might want to pursue further?

**Comment by michihdeu** [comment]
> Still anything in this issue we might want to pursue further?

Yep, the initial issue still exists:

> we have a lot of unmatched fps in unmatchedfps.log. It would be good to **automate a one-time cleanup** of the fp list.

We have 328 unmatched FPs right now: https://travelmapping.net/logs/unmatchedfps.log
